Add support for webpack-dev-server

renchap opened the pull request:
As discussed in #52, here is a proposal.
Documentation needs to be done, but I wanted some feedback first.

This is the simplest implementation possible I think. An alternative to install `webpack-dev-server` when installing Webpacker would be to add a check in `bin/webpack-dev-server` and ask the user to add it to `vendor/package.json` and run Yarn.
Another solution would be to not install the binstub, and create a new `webpack::install::dev_server` generator.

Note: this does not enable Hot Module Reloading or anything special, it only serves the packs using `webpack-dev-server` running on port 8080 by default. Adding HMR requires changes to your Webpack config (and changes to your application code / React integration / ... to be really useful) and I would prefer the user or a plugin to do it (I plan to add this to my React integration gem).

Any thoughts ?

guilleiguaran commented on lib/install/bin/webpack-dev-server.tt:
```diff
@@ -0,0 +1,15 @@
+WEBPACK_CONFIG = "#{APP_PATH}/config/webpack/#{WEBPACK_ENV}.js"
+
+Dir.chdir(VENDOR_PATH) do
+	exec "#{SET_NODE_PATH} #{WEBPACKER_BIN} --config #{WEBPACK_CONFIG} --content-base public/packs #{ARGV.join(" ")}"
```
extra indentation here

guilleiguaran commented on lib/install/config/shared.js:
```diff
@@ -13,7 +13,7 @@ module.exports = {
   ),
 
-  output: { filename: '[name].js', path: path.join('..', 'public', 'packs') },
+  output: { filename: '[name].js', path: path.resolve('..', 'public', 'packs') },
```
I think this change is already done in master

dhh commented on lib/webpacker/source.rb:
```diff
@@ -29,4 +31,8 @@ def digested_filename
+
+    def dev_server?
+      Rails.configuration.x.webpacker[:dev_server] && Rails.configuration.x.webpacker[:dev_server] != false
+    end
```
Instead of this, I'd like to add an accessor method like this:

```ruby
def config
  Rails.configuration.x.webpacker
end
```

Which we can then use like:

```ruby
def path
  if config[:dev_host].present?
    "#{config[:dev_host]}/#{filename}"
  elsif digesting?

  ...
end
```

That'll follow the pattern from `asset_host` we already use.

renchap commented:
Rebased to master and fixed your comments.

Do you think there should be a check in the binstub to point to the configuration if `Rails.configuration.x.webpacker[:dev_server]` is not set? What would be the most correct way to do this?

dhh commented on lib/webpacker/source.rb:
```diff
@@ -8,7 +8,9 @@ def initialize(name)
 
   def path
-    if digesting?
+    if config[:dev_server].present?
```
Let's use `dev_host` to match `asset_host`

renchap replied:
`dev_server_host` to make it clear this is about `webpack-dev-server` ?

dhh commented:
I think that would be a kind thing to do, but it would require booting all of Rails to provide that check. Not sure that's worth the trade-off.

dhh commented:
That works!

On Thu, Jan 12, 2017 at 3:45 PM, Renaud Chaput <notifications@github.com>
wrote:

> *@renchap* commented on this pull request.
> ------------------------------
>
> In lib/webpacker/source.rb <https://github.com/rails/webpacker/pull/55>:
>
> > @@ -8,7 +8,9 @@ def initialize(name)
>    end
>
>    def path
> -    if digesting?
> +    if config[:dev_server].present?
>
> dev_server_host to make it clear this is about webpack-dev-server ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rails/webpacker/pull/55>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAKtX0i7NZUpVK8JADGbNXYBhg1AahKks5rRjxygaJpZM4LgBKa>
> .
>

renchap commented:
Changed the config option to `dev_server_host` and added some documentation to README.

For the configuration check, I was thinking about reading `config/environment/#{RAILS_ENV}.rb` for something like `/[^#].*config\.x\.webpacker\[\:dev_server_host\].*=/`. Though I dont know if you want this kind of regexp check and this will not work if this is set elsewhere. An alternative would be to always print an information message before launching `webpack-dev-server`.

I dont have any preference here, I will let you decide what is the best.

dhh commented:
Great. I think the regexp check is fine for now 👍

renchap commented:
Here you go!

As a side note, `environment` in a generator inserts at the top of the environment file, just after `Rails.application.configure do`. Is there a reason for not appending at the end of the block? It would probably make the environment file look cleaner with base Rails options at first, then the ones from gems.

guilleiguaran commented:
@renchap thank you very much for this!!!

I agree we need to check the `environment` method in generators on Rails side.
